language/move-prover/src/bytecode_translator.rs

```diff
@@ -0,0 +1,1440 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module translates the bytecode of a module to Boogie code.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use itertools::Itertools;
+#[allow(unused_imports)]
+use log::{debug, info};
+
+use spec_lang::env::{FunctionEnv, GlobalEnv, Loc, ModuleEnv, Parameter, StructEnv, TypeParameter};
+use spec_lang::ty::{PrimitiveType, Type};
+use stackless_bytecode_generator::{
+    stackless_bytecode::StacklessBytecode::{self, *},
+    stackless_bytecode_generator::{StacklessFunction, StacklessModuleGenerator},
+};
+
+use crate::cli::Options;
+use crate::spec_translator::SpecTranslator;
+use crate::{
+    boogie_helpers::{
+        boogie_field_name, boogie_function_name, boogie_invariant_target, boogie_local_type,
+        boogie_struct_name, boogie_struct_type_value, boogie_type_check, boogie_type_value,
+        boogie_type_values, boogie_var_before_borrow,
+    },
+    cli::InvariantModel,
+    code_writer::CodeWriter,
+};
+use bytecode_to_boogie::lifetime_analysis::LifetimeAnalysis;
+use bytecode_to_boogie::stackless_control_flow_graph::StacklessControlFlowGraph;
+use num::Zero;
+
+pub struct BoogieTranslator<'env> {
+    env: &'env GlobalEnv,
+    writer: &'env CodeWriter,
+    options: &'env Options,
+}
+
+pub struct ModuleTranslator<'env> {
+    writer: &'env CodeWriter,
+    options: &'env Options,
+    module_env: ModuleEnv<'env>,
+    stackless_bytecode: Vec<StacklessFunction>,
+}
+
+impl<'env> BoogieTranslator<'env> {
+    pub fn new(env: &'env GlobalEnv, options: &'env Options, writer: &'env CodeWriter) -> Self {
+        Self {
+            env,
+            writer,
+            options,
+        }
+    }
+
+    pub fn translate(&mut self) {
+        // generate definitions for all modules.
+        for module_env in self.env.get_modules() {
+            ModuleTranslator::new(self, module_env).translate();
+        }
+    }
+}
+
+impl<'env> ModuleTranslator<'env> {
+    /// Creates a new module translator. Calls the stackless bytecode generator and wraps
+    /// result into the translator.
+    fn new(parent: &'env BoogieTranslator, module: ModuleEnv<'env>) -> Self {
+        let stackless_bytecode =
+            StacklessModuleGenerator::new(module.get_verified_module()).generate_module();
+        Self {
+            writer: parent.writer,
+            options: parent.options,
+            module_env: module,
+            stackless_bytecode,
+        }
+    }
+
+    /// Returns true if for the module no code should be produced because its already defined
+    /// in the prelude.
+    pub fn is_module_provided_by_prelude(&self) -> bool {
+        let name = self.module_env.get_name();
+        self.module_env.symbol_pool().string(name.name()).as_str() == "Vector"
+            && name.addr().is_zero()
+    }
```

Comment:
At some point, we may want to allow a mixture of functions, some from the prelude and some from Move code in the module.  The old vector.mvir had some native functions and some Move functions.
Also, we will sometimes want to have a hand-written Boogie function replace a Move function.
None of this needs to be worried about now.

Reply:
Agreed. We are hitting this already for Vector, some functions there are native others not, but in order to have an optimal Boogie representation, we skip all of them and need to handcode every one. Some flag per function would be more flexible.